numpyro/contrib/control_flow/scan.py

```diff
@@ -78,7 +83,99 @@ def _subs_wrapper(subs_map, i, length, site):
                                f" but got {value_ndim}. Please report the issue to us!")
 
 
-def scan_wrapper(f, init, xs, length, reverse, rng_key=None, substitute_stack=[]):
+class promote_shapes(Messenger):
+    # a helper messenger to promote shapes of `fn` and `value`
+    def process_message(self, msg):
+        if msg["type"] == "sample" and msg["value"] is not None:
+            fn, value = msg["fn"], msg["value"]
+            value_batch_ndims = jnp.ndim(value) - fn.event_dim
+            fn_batch_ndim = len(fn.batch_shape)
+            prepend_shapes = (1,) * abs(fn_batch_ndim - value_batch_ndims)
+            if fn_batch_ndim > value_batch_ndims:
+                msg["value"] = jnp.reshape(value, prepend_shapes + jnp.shape(value))
+            elif fn_batch_ndim < value_batch_ndims:
+                msg["fn"] = tree_map(lambda x: jnp.reshape(x, prepend_shapes + jnp.shape(x)), fn)
+
+
+def scan_enum(f, init, xs, length, reverse, rng_key=None, substitute_stack=None):
+    # XXX: This implementation only works for history size=1 but can be
+    # extended to history size > 1 by running `f` `history_size` times
+    # for initialization. However, `sequential_sum_product` does not
+    # support history size > 1, so we skip supporting it here.
```

Comment:
Note `funsor.sum_product.sarkka_bilmes_product` does support `history > 1`.

Reply:
Oh, I didn't know about that. Nice, thanks for pointing out! I temp into trying it this weekend. :D